tools/tinygobb/builder/builder.go

```diff
@@ -0,0 +1,328 @@
+func (b *Builder) AddJob(job BuildJob) error {
+	if b.state != Setup {
+		return fmt.Errorf("job: builder in invalid state, cannot add new job")
+	}
+
+	if job.goPkgPath == "" {
+		return fmt.Errorf("job: invalid gopkg, cannot add '%v'", job.goPkgPath)
+	}
+
+	b.jobs = append(b.jobs, job)
+	return nil
+}
+
+// Provide a custom logger for the builder. By default, the logger will be turned off.
+func (b *Builder) SetLogger(logger *log.Logger) {
+	b.logger = logger
+}
+
+// The worker function that will be run by the Builder.
+// The job queue as well as the result and error queue are provided by the builder.
+// TODO: make the channels unidirectional?
+// TODO: how about the mutex syncing for the logger? maybe just leave it out for now
+// func worker(jobq chan BuildJob, resultq chan BuildResult, errq chan BuildError, id uint, logger *log.Logger) {
+func worker(jobq chan BuildJob, resultq chan BuildResult, errq chan BuildError, id uint) {
+	fmt.Printf("[%d] spawned worker %d\n", id, id)
+	for {
+		job, ok := <-jobq
+		if !ok {
+			fmt.Printf("[%d] finish worker %d\n", id, id)
+			return
+		}
+
+		var errBuf bytes.Buffer
+
+		// run the build process
+		fields := strings.Fields(job.buildCommand)
+		fields = append(fields, "-o", *job.buildDir)
+
+		c := exec.Command(fields[0], fields[1:]...)
+		c.Env = append(os.Environ(), job.env...)
+		c.Dir = job.goPkgPath
+		c.Stderr = &errBuf
+
+		fmt.Printf("[%d] running '%v'\n", id, c)
+		t0 := time.Now()
+		err := c.Run()
+		t1 := time.Now()
+
+		if err != nil {
+			// the subprocess failed with a non-zero exit code, so create BuildError
+			fmt.Printf("[%d] error building %v:%v", id, job.goPkgPath, errBuf.String())
+			errq <- BuildError{
+				buildJob: job,
+				buildErr: errBuf.String(),
+			}
+		} else {
+			fmt.Printf("[%d] built %v\n", id, job.goPkgPath)
+			pkgNameToken := strings.Split(job.goPkgPath, "/")
+			binPath := job.goPkgPath + "/" + pkgNameToken[len(pkgNameToken)-1]
+
+			f, err := os.Stat(binPath)
+			if err != nil {
+				fmt.Printf("worker: could not find file %v\n", binPath)
+			}
+
+			resultq <- BuildResult{
+				buildJob:   job,
+				buildTime:  t1.Sub(t0),
+				binarySize: uint64(f.Size()),
+			}
+		}
+	}
+}
+
+// Start execution of the builder. The builder will be set into the Running state
+// and will block all methods modifying the Builders stats.
+func (b *Builder) Run() error {
+	if b.state != Setup {
+		return ErrBuilderInvalidState
+	}
+	b.state = Running
+
+	if len(b.jobs) == 0 {
+		return ErrBuilderNoJobs
+	}
+
+	nJobs := len(b.jobs)
+
+	// spawn the workers
+	for id := range b.worker {
+		b.logger.Printf("spawning %d workers", b.worker)
+		// go worker(b.jobQueue, b.resultQueue, b.errQueue, id, b.logger)
+		go worker(b.jobQueue, b.resultQueue, b.errQueue, id)
+	}
+
+	// make sure all the workers are spawned and ready.
+	// TODO: make this smarter with a wg or a callback channel
+	time.Sleep(time.Second)
```

Comment:
ouch. Please don't do this Sleep :-) I've started using errgroup and it's pretty nice, consider using it.